iodata/formats/mwfn.py

```diff
@@ -0,0 +1,315 @@
+        "mo_sym": np.empty(n_mo, str),
+        "mo_coeffs": np.empty([n_basis, n_mo], float),
+    }
+
+    for index in range(n_mo):
+        line = next(lit)
+        while 'Index' not in line:
+            line = next(lit)
+        assert line.startswith('Index')
+        data["mo_numbers"][index] = int(line.split()[1])
+        data["mo_type"][index] = int(next(lit).split()[1])
+        data["mo_energies"][index] = float(next(lit).split()[1])
+        data["mo_occs"][index] = float(next(lit).split()[1])
+        data["mo_sym"][index] = str(next(lit).split()[1])
+        # skip "$Coeff line
+        next(lit)
+        data["mo_coeffs"][:, index] = _load_helper_section(lit, n_basis, '', 0, float)
+
+    return data
+
+
+def _load_mwfn_low(lit: LineIterator) -> dict:
+    """Load data from a MWFN file into arrays.
+
+    Parameters
+    ----------
+    lit
+        The line iterator to read the data from.
+    """
+    # Note:
+    # -----
+    # 1) mwfn is a fortran program which loads *.mwfn by locating the line with the keyword,
+    #    then uses `backspace`, then begins reading. Despite this flexibility, it is stated by
+    #    the authors that the order of section, and indeed, entries in general, must be fixed.
+    #    With this in mind the input utilized some hard-coding since order should be fixed.
+    # 2) mwfn ignores lines beginning with `#`.
+
+    # read title (assuming title is the 1st line, which seems to be the standard)
+    data = {"title": next(lit).strip()}
+
+    # load Wfntype, Charge, Naelec, Nbelec, E_tot, VT_ratio, & Ncenter
+    data.update(_load_helper_opener(lit))
+
+    # load atnums, atcorenums, & atcoords (in atomic units)
+    data.update(_load_helper_atoms(lit, data["Ncenter"]))
+
+    # load Nbasis, Nindbasis, Nprims, Nshell, & Nprimshell
+    data.update(_load_helper_basis(lit))
+
+    # load shell_types, shell_centers, & shell_contraction_degrees
+    data.update(_load_helper_shells(lit, data["Nshell"]))
+    # IOData indices start at 0, so the centers are shifted
+    data["shell_centers"] -= 1
+
+    # load primitive exponents & coefficients
+    if not next(lit).startswith("$Primitive exponents"):
+        lit.error("Expected '$Primitive exponents' section!")
+    data["exponents"] = _load_helper_section(lit, data["Nprimshell"], '', 0, float)
+    if not next(lit).startswith("$Contraction coefficients"):
+        lit.error("Expected '$Contraction coefficients' section!")
+    data["coeffs"] = _load_helper_section(lit, data["Nprimshell"], '', 0, float)
+
+    # get number of basis & molecular orbitals (MO)
+    # Note: MWFN includes virtual orbitals, so num_mo equals number independent basis functions
+    num_basis = data["Nindbasis"]
+    num_mo = data["Nindbasis"]
+    if data["mo_kind"] == "unrestricted":
+        num_mo *= 2
+    # load MO information
+    data.update(_load_helper_mo(lit, num_basis, num_mo))
+
+    # TODO: load data in '# Various matrices' section
```

Comment:
Should we open an issue for this? I'm worried that leaving it here as a TODO item will not be noticed.

Reply:
I had the same TODO, but worded differently. The issue I had was that the file is read line by line, and so when scanning for the optional keywords, I would hit end of file and get an error, if they weren't there. I think the fix is probably simple. A brute force option would be to open the file, scan or optional keywords, then proceed with normal scan. An another would be to have the optional keywords passed by caller. I feel however, that there is a simple fix to the end-of-line error.

Reply:
Alternatively, to avoid searching for optional keywords (and going through the file twice), we can have a list of all possible keywords that appear in `mwfn` files, and use a `while` loop to properly parse what is in a file (like 'fchk')... This would be more robust and easier to extend. 

Currently we are relying a specific order of data in `mwfn` file, which would not be problematic as only `Multiwfn` writes `mwfn` files, and as @BradenDKelly commented the order of entries must be fixed. The `'# Various matrices'` section includes alpha/beta/total density matrix, overlap matrix, etc. (we can get a list of these from the paper but it is meant to be a place for dumping matrices used for post processing). As this PR is getting long, I think it's best to open an issue for paring `'# Various matrices'` section of mwfn @tovrstra.

Reply:
I'll make the issue and remove the TODO line, because it may trip pylint (depends on local config).